dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -233,23 +255,92 @@ class AsyncTunnelSender : public TunnelSender
     GRPCSendQueue<TrackedMppDataPacketPtr> queue;
 };
 
-/// LocalTunnelSender just provide readForLocal method to return one element one time
-/// LocalTunnelSender is owned by the associated ExchangeReceiver
+template <bool enable_fine_grained_shuffle>
 class LocalTunnelSender : public TunnelSender
 {
 public:
-    using Base = TunnelSender;
-    using Base::Base;
-    TrackedMppDataPacketPtr readForLocal();
+    LocalTunnelSender(
+        size_t source_index_,
+        const String & req_info_,
+        SupportForLocalExchange & support_for_local_,
+        const LoggerPtr & log_,
+        MemoryTrackerPtr & memory_tracker_,
+        const String & tunnel_id_)
+        : TunnelSender(memory_tracker_, log_, tunnel_id_, nullptr)
+        , source_index(source_index_)
+        , req_info(req_info_)
+        , support_for_local(support_for_local_)
+        , is_done(false)
+    {}
+
+    ~LocalTunnelSender() override
+    {
+        closeLocalTunnel(false, "");
+    }
+
+    bool push(TrackedMppDataPacketPtr && data) override
+    {
+        if (unlikely(checkPacketErr(data)))
```

Comment:
> According to original implementation, even if `data` contains error message, it is send to the receiver side, because it is the way to pass the error message from one task to another.

In new implementation, `checkPacketErr` can also pass err message to receiver task by calling `closeLocalTunnel` function.

Reply:
But the read thread(`ExchangeReceiverBase<RPCContext>::nextResult`) does not check error message, if you want to use the new implementation, you need to refind the read logical in `ExchangeReceiver`